lib/engine/game/g_18_al.rb

```diff
@@ -46,11 +54,86 @@ def operating_round(round_num)
         ], round_num: round_num)
       end
 
+      def stock_round
+        Round::Stock.new(self, [
+          Step::DiscardTrain,
+          Step::G18AL::BuySellParShares,
+        ])
+      end
+
+      def action_processed(action)
+        super
+        case action
+        when Action::Assign
+          company = action.entity
+          target = action.target
+
+          # Add a revenue bonus for the corporation
+          # that will be removed in phase 6
+          ability = Engine::Ability::HexesBonus.new(
+            type: :hexes_bonus,
+            name: 'Coal Field token',
+            display_value: target.id,
+            **{
+              'hexes' => [target.id],
+              'amount' => 10,
+            }
+          )
+          company.owner.add_ability(ability)
+          @hexes.select { |hex| %w[H3 G4 H5 G6 E6].include?(hex.name) }
```

Comment:
you're creating this array multiple times, you should make it a constant

Reply:
Done